<a href="https://colab.research.google.com/github/summayahjoyia/dico-nli-starter/blob/main/notebooks/dico_nli_work.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import os
from google.colab import userdata

token = userdata.get('GITHUB_TOKEN')
user = "summayahjoyia"
repo = "dico-nli-starter"

if not os.path.exists(f"/content/{repo}"):
    !git clone -q https://{user}:{token}@github.com/{user}/{repo}.git /content/{repo}

%cd /content/{repo}
!git config user.name "Summayah Javed"
!git config user.email "summayah.joyia@gmail.com"
!git status

/content/dico-nli-starter
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


Downloading the task data and scorer

In [3]:
%cd /content
!git clone -q https://github.com/ilopezgazpio/SemEval-2027-Task-2-DiCo-NLI /content/task_repo
!ls /content/task_repo
!ls /content/task_repo/final_data
!ls /content/task_repo/starter_kit

/content
advertising  evaluation_functions  LICENSE     starter_kit
assets	     final_data		   README.md   static
_config.yml  latex		   References  trial_data
dev  final_data_dev.zip  final_data_train.zip  README.md  train
data  main.py  README.md  requirements.txt  results  SLURM  src


Loading the data

In [4]:
!ls /content/task_repo/final_data/train
!ls /content/task_repo/final_data/dev

import pandas as pd
train = pd.read_csv("/content/task_repo/final_data/train/dico_nli_train_track1_participant_labeled.csv")
dev = pd.read_csv("/content/task_repo/final_data/dev/dico_nli_dev_track1_reference.csv")
print("TRAIN columns:", list(train.columns), "| rows:", len(train))
print("DEV columns:", list(dev.columns), "| rows:", len(dev))
print(train.head(3))
print(dev.head(3))

dico_nli_train_track1_participant_labeled.csv
dico_nli_train_track1_reference.csv
dico_nli_train_track1_submission_template.csv
dico_nli_train_track2_participant_labeled.csv
dico_nli_train_track2_reference.csv
dico_nli_train_track2_submission_template.csv
dico_nli_train_track3_participant_labeled.csv
dico_nli_train_track3_reference.csv
dico_nli_train_track3_submission_template.csv
dico_nli_train_track4_participant_labeled.csv
dico_nli_train_track4_reference.csv
dico_nli_train_track4_submission_template.csv
dico_nli_dev_track1_participant_labeled.csv
dico_nli_dev_track1_reference.csv
dico_nli_dev_track1_submission_template.csv
dico_nli_dev_track2_participant_labeled.csv
dico_nli_dev_track2_reference.csv
dico_nli_dev_track2_submission_template.csv
dico_nli_dev_track3_participant_labeled.csv
dico_nli_dev_track3_reference.csv
dico_nli_dev_track3_submission_template.csv
dico_nli_dev_track4_participant_labeled.csv
dico_nli_dev_track4_reference.csv
dico_nli_dev_track4_submission_template.csv


Finds the most frequent label in train and predicts in for every dev item.

In [5]:
import os
os.makedirs("/content/task_repo/results", exist_ok=True)
#Return the label with the biggest count after reading all the labels and their counts
majority = train["label"].value_counts().idxmax()
print("Majority label:", majority)
print(train["label"].value_counts())
#Building a new table with columns of dev and the majority label
pred = pd.DataFrame({"instance_id": dev["instance_id"], "label": majority})
pred.to_csv("/content/task_repo/results/majority_dev.csv", index=False)
print(pred.head(), len(pred))

Majority label: FORWARD_ENTAILMENT
label
FORWARD_ENTAILMENT     881
BACKWARD_ENTAILMENT    881
EQUIVALENCE            802
NEGATIVE_OTHER         478
Name: count, dtype: int64
                             instance_id               label
0   dico_phrasis_0000002__en-en__flipped  FORWARD_ENTAILMENT
1  dico_phrasis_0000002__en-en__original  FORWARD_ENTAILMENT
2   dico_phrasis_0000004__en-en__flipped  FORWARD_ENTAILMENT
3  dico_phrasis_0000004__en-en__original  FORWARD_ENTAILMENT
4   dico_phrasis_0000008__en-en__flipped  FORWARD_ENTAILMENT 660


Running the official scorer.

In [6]:
%cd /content/task_repo
!python3 -m evaluation_functions --gold final_data/dev/dico_nli_dev_track1_reference.csv --predictions results/majority_dev.csv --output-dir results/majority_scores
!ls results/majority_scores

/content/task_repo
{
  "classification": {
    "accuracy": 0.2878787878787879,
    "confusion_matrix": {
      "BACKWARD_ENTAILMENT": {
        "BACKWARD_ENTAILMENT": 0,
        "EQUIVALENCE": 0,
        "FORWARD_ENTAILMENT": 190,
        "NEGATIVE_OTHER": 0
      },
      "EQUIVALENCE": {
        "BACKWARD_ENTAILMENT": 0,
        "EQUIVALENCE": 0,
        "FORWARD_ENTAILMENT": 174,
        "NEGATIVE_OTHER": 0
      },
      "FORWARD_ENTAILMENT": {
        "BACKWARD_ENTAILMENT": 0,
        "EQUIVALENCE": 0,
        "FORWARD_ENTAILMENT": 190,
        "NEGATIVE_OTHER": 0
      },
      "NEGATIVE_OTHER": {
        "BACKWARD_ENTAILMENT": 0,
        "EQUIVALENCE": 0,
        "FORWARD_ENTAILMENT": 106,
        "NEGATIVE_OTHER": 0
      }
    },
    "macro_f1": 0.11176470588235295,
    "per_label": {
      "BACKWARD_ENTAILMENT": {
        "f1": 0.0,
        "false_negative": 190,
        "false_positive": 0,
        "label": "BACKWARD_ENTAILMENT",
        "precision": 0.0,
        "predicted"

In [7]:
%cd /content/dico-nli-starter
!git add -A
!git commit -m "Part 0: trivial baseline predictions and scores"
!git pull --rebase
!git push

/content/dico-nli-starter
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
Already up to date.
Everything up-to-date


In [8]:
%cd /content/dico-nli-starter
!git pull --rebase
!git push

/content/dico-nli-starter
Already up to date.
Everything up-to-date


In [9]:
!git -C /content/task_repo rev-parse HEAD

588968e610197ddc4c440314701cbc587afc4c1b


Labelling distributions (Part 1)

In [16]:
import os
from IPython.display import display

def dist(df):
    c = df["label"].value_counts()
    return pd.DataFrame({"count": c, "percent": (c / len(df) * 100).round(1)})

t1 = dist(train).join(dist(dev), lsuffix="_train", rsuffix="_dev")
t1.loc["TOTAL"] = [len(train), 100.0, len(dev), 100.0]
t1[["count_train", "count_dev"]] = t1[["count_train", "count_dev"]].astype(int)

display(t1)

,count_train,percent_train,count_dev,percent_dev
label,,,,
FORWARD_ENTAILMENT,881,29.0,190,28.8
BACKWARD_ENTAILMENT,881,29.0,190,28.8
EQUIVALENCE,802,26.4,174,26.4
NEGATIVE_OTHER,478,15.7,106,16.1
TOTAL,3042,100.0,660,100.0


Three examples of each label, with a reason (Part 1)


In [11]:
from IPython.display import display

#skipping out on any flipped statements for now
orig = train[train["instance_id"].str.endswith("__original")]
# printing 3 random pairs per label
table= (orig.groupby("label").sample(3,random_state=2)[["label","text1","text2"]].sort_values("label").reset_index(drop=True))
table["why"]=["Two small birds implies two birds but two birds does not imply that they are small specifically",
              "Israel planes does not imply that there are israel airstrikes but israel airstrikes does mean that there were israel planes involved",
              "A red double decker bus is a double decker bus so it's true in the backward direction but a double decker bus is not necessarily red thus false in forward direction",
              "A suspected US drone strike and a US drone strike describe the same event, and 'suspected' only says how sure we are about it, so they are labeled equivalent, though strictly a suspected strike might not have actually happened",
              "The Philippine nation implies the Philippines as a country and if we say the Philippines, it can also be called the Philippine nation",
              "Both texts mean the same thing, they only differ in the hyphen and the spelling (Philips vs Phillips), thus equivalent",
              "Being on a city street means in a city but in a city does not always mean on a city street so its only true in the forward direction",
              "text1 specifies which protesters and text2 is general protesters so the specific kind of protesters are a subgroup of general protesters but general protesters are not contained within a specific kind, so only true in the forward direction"
,"I can't justify this label because text1 only mentions black and white cows and text2 brings in a brown cow that text1 never mentions, so text1 does not entail text2 and the label looks questionable",
              "Going into the water does not mean you're under water, you could be at the surface, and being underwater does not mean you went into the water, you could have been there all along, they are both related but one is a motion and the other is a state so neither entails the other",
              "I believe this one should be forward entailment because all text1 is adding is the detail of 'record setting' so text1 does imply text2 however it has been labelled NEGATIVE_OTHER, this could be due to the grammar detail of 'in' in text2",
              "losing contact does not imply that something vanished, vanishing could possibly imply contact is lost but not necessarily, they're similar and could be appearing in the same news but they are not the same event",
              ]
print("Examples")

display(table)




Examples


,label,text1,text2,why
0,BACKWARD_ENTAILMENT,Two birds,Two small birds,Two small birds implies two birds but two bird...
1,BACKWARD_ENTAILMENT,Israel planes,Israel airstrikes,Israel planes does not imply that there are is...
2,BACKWARD_ENTAILMENT,A double decker bus,A red double decker bus,A red double decker bus is a double decker bus...
3,EQUIVALENCE,Suspected U.S. drone strike,US drone strike,A suspected US drone strike and a US drone str...
4,EQUIVALENCE,the Philippine nation,Philippines,The Philippine nation implies the Philippines ...
5,EQUIVALENCE,A Philips flat-screen TV,A Phillips flat screen TV,"Both texts mean the same thing, they only diff..."
6,FORWARD_ENTAILMENT,on a city street,in a city,Being on a city street means in a city but in ...
7,FORWARD_ENTAILMENT,at pro-Morsy protesters,at protesters,text1 specifies which protesters and text2 is ...
8,FORWARD_ENTAILMENT,The black and white cows,A black cow and a brown cow,I can't justify this label because text1 only ...
9,NEGATIVE_OTHER,into the water,underwater,Going into the water does not mean you're unde...


How reversed pairs are linked in the data (Part 1)

In [12]:
#Checking one example
row = dev.iloc[0]
twin = dev[dev["instance_id"]== row["reverse_pair_id"]].iloc[0]
print(row[["instance_id", "pair_id", "reverse_pair_id", "text1", "text2", "label"]].to_string())
print()
print(twin[["instance_id", "pair_id", "reverse_pair_id", "text1", "text2", "label"]].to_string())

instance_id         dico_phrasis_0000002__en-en__flipped
pair_id                             dico_phrasis_0000002
reverse_pair_id    dico_phrasis_0000002__en-en__original
text1                                's war shrine visit
text2                                    's shrine visit
label                                 FORWARD_ENTAILMENT

instance_id        dico_phrasis_0000002__en-en__original
pair_id                             dico_phrasis_0000002
reverse_pair_id     dico_phrasis_0000002__en-en__flipped
text1                                    's shrine visit
text2                                's war shrine visit
label                                BACKWARD_ENTAILMENT


In [13]:
#Checking if this holds for the entire dataset
ids = set(dev["instance_id"])
print("All dev reverse_pair_ids exist:", dev["reverse_pair_id"].isin(ids).all())
print("Dev rows with missing reverse_pair_id:", dev["reverse_pair_id"].isna().sum())


All dev reverse_pair_ids exist: False
Dev rows with missing reverse_pair_id: 106


In [14]:
missing = set(dev[dev["reverse_pair_id"].isna()]["instance_id"])
neg_other = set(dev[dev["label"] == "NEGATIVE_OTHER"]["instance_id"])

print("Missing reverse_pair_id:", len(missing))
print("NEGATIVE_OTHER:", len(neg_other))
print("Same rows:", missing == neg_other)
print("In missing but not NEGATIVE_OTHER:", len(missing - neg_other))
print("In NEGATIVE_OTHER but not missing:", len(neg_other - missing))

Missing reverse_pair_id: 106
NEGATIVE_OTHER: 106
Same rows: True
In missing but not NEGATIVE_OTHER: 0
In NEGATIVE_OTHER but not missing: 0


Each reversed pair is stored as two rows whose instance_id is share pair_id and end in __original or __flipped. In the dev reference file, the reverse_pair_id column holds the instance_id of the other row, so each twin points to the other. The twin has text1 and text2 swapped and its label reversed (FORWARD and BACKWARD are swapped but EQUIVALENCE stays the same). NEGATIVE_OTHER have no reverse(their reverse_pair_id is empty, 106 of 660 rows as shown in above code cell.)

Pushing the examples table to GitHub

In [15]:
import os
out = "/content/dico-nli-starter/results/part1"
os.makedirs(out, exist_ok=True)

table.to_csv(f"{out}/examples.csv", index=False)

%cd /content/dico-nli-starter
!git add -A
!git commit -m "Part 1: data exploration examples"
!git pull --rebase
!git push

/content/dico-nli-starter
[main 1f6b3be] Part 1: data exploration examples
 1 file changed, 13 insertions(+)
 create mode 100644 results/part1/examples.csv
Current branch main is up to date.
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 2 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (5/5), 1.50 KiB | 1.50 MiB/s, done.
Total 5 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/summayahjoyia/dico-nli-starter.git
   9c8a91d..1f6b3be  main -> main


In [18]:
import os
out = "/content/dico-nli-starter/results/part1"
os.makedirs(out, exist_ok=True)
t1.to_csv(f"{out}/label_distribution.csv")

In [20]:
%cd /content/dico-nli-starter
!git add -A
!git commit -m "Part 1: label distribution table"
!git pull --rebase
!git push

/content/dico-nli-starter
[main de27a40] Part 1: label distribution table
 1 file changed, 6 insertions(+)
 create mode 100644 results/part1/label_distribution.csv
Current branch main is up to date.
Enumerating objects: 8, done.
Counting objects: 100% (8/8), done.
Delta compression using up to 2 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 602 bytes | 602.00 KiB/s, done.
Total 5 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/summayahjoyia/dico-nli-starter.git
   1f6b3be..de27a40  main -> main
